test/integration/filters/set_route_filter.cc

```diff
@@ -0,0 +1,66 @@
+#include <string>
+
+#include "envoy/http/filter.h"
+#include "envoy/registry/registry.h"
+#include "envoy/server/filter_config.h"
+
+#include "common/router/delegating_route_impl.h"
+
+#include "extensions/filters/http/common/pass_through_filter.h"
+
+#include "test/extensions/filters/http/common/empty_http_filter_config.h"
+#include "test/integration/filters/common.h"
+
+namespace Envoy {
+
+// A test filter that sets the cached route via setRoute callback and uses the
+// DelegatingRoute mechanism to override the finalized upstream cluster.
+class SetRouteFilter : public Http::PassThroughFilter {
+public:
+  constexpr static char name[] = "set-route-filter";
+
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap&, bool) override {
+    std::shared_ptr<const ExampleDerivedDelegatingRoute> route_override =
+        std::make_shared<ExampleDerivedDelegatingRoute>(decoder_callbacks_->route(),
+                                                        "cluster_override");
+
+    decoder_callbacks_->setRoute(route_override);
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  // TODO: Put in one place to share with unit tests (test/common/http:conn_manager_impl_test)
```

Comment:
would be easy enough to add this to a file in /test/common/http to be reusable, maybe do that in this PR?

Reply:
done